<table border=0 width="100%"><tr><td><p align="left"><img src="Images/logo.png" align="left" width=300></p></td><td><font size=5><B>Python程序设计与应用</B></font></td></tr></table>

# <div align='center'>第六部分 字典数据类型</div> 

## 知识点
* 字典数据类型
* 字典类型的操作

## 预习
* 第六章：组合数据类型

# 1. 字典数据类型
字典数据(dictionary)类型提供了一种灵活的访问和组织数据的方式。像列表一样，“字典”是许多值的集合。但不像列表的下标，***字典的索引可以使用许多不同数据类型***，不只是整数。字典的索引被称为“键”，键及其关联的值称为“***键-值***”对。

Python语言中的字典可以通过大括号(**{}**)建立，建立模式如下：

<center> **{<键1>:<值1>, <键2>:<值2>, … , <键n>:<值n>} ** </center>

其中，键和值通过冒号连接，不同键值对通过逗号隔开。

例如，把下表中的一条雇员数据存储在字典employee中：

|编号|姓名|部门|电子邮箱|电话号码|身份证号码|
|---|---|---|---|---|---|
|10932|张珊|管理|zhans@163.com|18612345678|520125197907167551|
|10933|李思|软件|lisi@163.com|18612345679|510110195512176250|
|10934|王武|财务|wangwu@example.com|18612345680|510102196510255826|

In [2]:
employee = {"编号":"10932","姓名":"张珊","部门":"管理","电子邮箱":"zhans@163.com","电话号码":"18612345678","身份证号码":"520125197907167551"}

在这个字典中，编号、姓名、部门、电子邮箱、电话号码、身份证号码称为“**键**”，10932、张珊、管理、zhans@163.com、18612345678、520125197907167551称为这些键分别对应的“**值**”。

键通常是字符串，但它还可以是Python中其他任意的**不可变类型**：布尔型、整型、浮点型、元组、字符串，以及其他一些不可变的类型。字典本身是可变的，因此可以增加、删除或修改其中的键值对。字典的键必须保证互不相同。

In [3]:
tmpDic = {True:100, 108:("西南财大","电子科大"),3.14:"Pi"}

上面这个字典没有任何实际意义，但是，***请指出它的键分别有哪些？值分别有哪些？***

字典中的表项是不排序的。字典打印出来的顺序与创建之初的顺序不同，这不是错误。字典是集合类型的延续，各个元素并没有顺序之分。如果一样，纯属巧合:)

In [5]:
print(employee)

{'编号': '10932', '姓名': '张珊', '部门': '管理', '电子邮箱': 'zhans@163.com', '电话号码': '18612345678', '身份证号码': '520125197907167551'}


* **访问字典值**：

字典最主要的用法是查找与特定键相对应的值，这通过索引符号来实现。 

一般来说，字典中键值对的访问模式如下，采用中括号格式：
       
<center>**<值> = <字典变量>[<键>] **</center>
       
字典中对某个键值的修改可以通过中括号的访问和赋值实现：

In [8]:
employee['姓名']

'张珊'

In [10]:
employee['部门'] = '人事'

In [11]:
print(employee)

{'编号': '10932', '姓名': '张珊', '部门': '人事', '电子邮箱': 'zhans@163.com', '电话号码': '18612345678', '身份证号码': '520125197907167551'}


尝试访问字典中不存在的键，将导致KeyError 出错信息。这很像列表的“越界”IndexError 出错信息。

In [12]:
employee['salary']

KeyError: 'salary'

因为字典是不排序的，所以不能像列表那样切片。

In [7]:
employee['编号':'电话号码']

TypeError: unhashable type: 'slice'

* **添加键值对**

通过中括号可以增加新的元素，例如雇员信息中，发现少了性别，可以通过下面的方式添加:

In [13]:
employee

{'姓名': '张珊',
 '电子邮箱': 'zhans@163.com',
 '电话号码': '18612345678',
 '编号': '10932',
 '身份证号码': '520125197907167551',
 '部门': '人事'}

In [14]:
employee['性别'] ='女'

In [15]:
employee

{'姓名': '张珊',
 '性别': '女',
 '电子邮箱': 'zhans@163.com',
 '电话号码': '18612345678',
 '编号': '10932',
 '身份证号码': '520125197907167551',
 '部门': '人事'}

* **删除键值对**

通过del删除字典中某一个键值对:

In [17]:
del employee['性别']

In [18]:
employee

{'姓名': '张珊',
 '电子邮箱': 'zhans@163.com',
 '电话号码': '18612345678',
 '编号': '10932',
 '身份证号码': '520125197907167551',
 '部门': '人事'}

# 2. 字典类型的操作

|函数和方法|描述|
|-----|:-----|
|<d>.keys()|返回所有的键信息|
|<d>.values()|返回所有的值信息||
|<d>.items()|返回所有的键值对|
|<d>.get(<key>,<default>)|键存在则返回相应值，否则返回默认值|
|<d>.pop(<key>,<default>)|键存在则返回相应值，同时删除键值对，否则返回默认值|
|<d>.popitem()|随机从字典中取出一个键值对，以元组(key, value)形式返回|
|<d>.clear()|删除所有的键值对|
|del <d>[<key>]|删除字典中某一个键值对|
|<key> in <d>|如果键在字典中返回True，否则返回False|

## 2.1 keys()、values()和items()方法
有3 个字典方法，它们将返回类似列表的值，分别对应于字典的键、值和键-值对：keys()、values()和items()。这些方法返回的值不是真正的列表，它们不能被修改，没有append()方法。但这些数据类型（分别是dict_keys、dict_values 和dict_items）可以用于for 循环。

In [19]:
for v in employee.values():
    print(v)

10932
张珊
人事
zhans@163.com
18612345678
520125197907167551


In [20]:
for k in employee.keys():
    print(k)

编号
姓名
部门
电子邮箱
电话号码
身份证号码


In [21]:
for i in employee.items():
    print(i)

('编号', '10932')
('姓名', '张珊')
('部门', '人事')
('电子邮箱', 'zhans@163.com')
('电话号码', '18612345678')
('身份证号码', '520125197907167551')


利用keys()、values()和items()方法，循环分别可以迭代键、值或键-值对。请注意，items()方法返回的dict_items 值中，包含的是键和值的元组。

如果希望通过这些方法得到一个真正的列表，就把类似列表的返回值传递给 list函数。

In [22]:
employee.keys()

dict_keys(['编号', '姓名', '部门', '电子邮箱', '电话号码', '身份证号码'])

In [23]:
list(employee.keys())

['编号', '姓名', '部门', '电子邮箱', '电话号码', '身份证号码']

也可以利用多重赋值的技巧，在for 循环中将键和值赋给不同的变量。

In [24]:
for k, v in employee.items():
    print("key:" + k + "  Value:" + v)

key:编号  Value:10932
key:姓名  Value:张珊
key:部门  Value:人事
key:电子邮箱  Value:zhans@163.com
key:电话号码  Value:18612345678
key:身份证号码  Value:520125197907167551


## 2.2 检查字典中是否存在键或值

前面提到，in 和not in 操作符可以检查值是否存在于列表中。也可以利用这些操作符，检查某个键或值是否存在于字典中。

In [25]:
'姓名' in employee

True

In [26]:
'张珊' in employee

False

In [27]:
'张珊' in employee.values()

True

In [28]:
'部门' in employee.keys()

True

In [29]:
'性别' not in employee

True

## 2.3 get()方法

在访问一个键的值之前，检查该键是否存在于字典中，这很麻烦。好在，字典有一个get()方法，它有两个参数：要取得其值的键，以及如果该键不存在时，返回的备用值。

In [31]:
curEmployeeSalary = employee['salary']

KeyError: 'salary'

In [32]:
if 'salary' in employee:
    curEmployeeSalary = employee['salary']
else:
    curEmployeeSalary = 0

print(curEmployeeSalary)

0


In [33]:
curEmployeeSalary = employee.get('salary',0)
print(curEmployeeSalary)

0


## 2.4 setdefault()方法

常常需要为字典中某个键设置一个默认值，当该键没有任何值时使用它。代码看起来像这样：

In [34]:
employee

{'姓名': '张珊',
 '电子邮箱': 'zhans@163.com',
 '电话号码': '18612345678',
 '编号': '10932',
 '身份证号码': '520125197907167551',
 '部门': '人事'}

In [35]:
if 'salary' not in employee:
    employee['salary'] = 0


In [36]:
employee

{'salary': 0,
 '姓名': '张珊',
 '电子邮箱': 'zhans@163.com',
 '电话号码': '18612345678',
 '编号': '10932',
 '身份证号码': '520125197907167551',
 '部门': '人事'}

setdefault()方法提供了一种方式，在一行中完成这件事。传递给该方法的第一个参数，是要检查的键。第二个参数，是如果该键不存在时要设置的值。

In [37]:
del employee['salary']

In [38]:
employee

{'姓名': '张珊',
 '电子邮箱': 'zhans@163.com',
 '电话号码': '18612345678',
 '编号': '10932',
 '身份证号码': '520125197907167551',
 '部门': '人事'}

In [39]:
employee.setdefault('salary', 0)

0

In [40]:
employee

{'salary': 0,
 '姓名': '张珊',
 '电子邮箱': 'zhans@163.com',
 '电话号码': '18612345678',
 '编号': '10932',
 '身份证号码': '520125197907167551',
 '部门': '人事'}

如果该键确实存在，方法就会返回键的值。

In [41]:
department = employee.setdefault('部门', '管理')

In [42]:
department

'人事'

**字典小实例：**

setdefault()方法是一个很好的快捷方式，可以确保一个键存在。下面有一个小程序，计算一个字符串中每个字符出现的次数。

In [44]:
message = "我爱程序设计，我要好好学习，天天向上。"
count = {}
for character in message:
    count.setdefault(character, 0)
    count[character] = count[character] + 1
    
print(count)

{'我': 2, '爱': 1, '程': 1, '序': 1, '设': 1, '计': 1, '，': 2, '要': 1, '好': 2, '学': 1, '习': 1, '天': 2, '向': 1, '上': 1, '。': 1}


## 2.5 使用update()合并字典

使用update() 可以将一个字典的键值对复制到另一个字典中去。

In [52]:
empInfo = {"编号":"10932","姓名":"张珊","部门":"管理","电子邮箱":"zhans@163.com","电话号码":"18612345678","身份证号码":"520125197907167551"}
empIncome = {"编号":"10932","电话号码":"18687654321","基本工资":15000.00,"奖金":4000.00,"住房补助":2300.00,"五险一金":980.00,"其他扣款":300.00 ,"专项附加扣除":1000.00}

In [53]:
empInfo

{'姓名': '张珊',
 '电子邮箱': 'zhans@163.com',
 '电话号码': '18612345678',
 '编号': '10932',
 '身份证号码': '520125197907167551',
 '部门': '管理'}

In [54]:
empIncome

{'专项附加扣除': 1000.0,
 '五险一金': 980.0,
 '住房补助': 2300.0,
 '其他扣款': 300.0,
 '基本工资': 15000.0,
 '奖金': 4000.0,
 '电话号码': '18687654321',
 '编号': '10932'}

In [55]:
empInfo.update(empIncome)

In [56]:
empInfo

{'专项附加扣除': 1000.0,
 '五险一金': 980.0,
 '住房补助': 2300.0,
 '其他扣款': 300.0,
 '基本工资': 15000.0,
 '奖金': 4000.0,
 '姓名': '张珊',
 '电子邮箱': 'zhans@163.com',
 '电话号码': '18687654321',
 '编号': '10932',
 '身份证号码': '520125197907167551',
 '部门': '管理'}

其中，

* 在字典empInfo和empIncome中都存在的键"编号"和"电话号码"，使用empIncome中的值替换；
* 只在字典empInfo中的键对应的值维持原值；
* 只在字典empIncome中的键对应的键值对被添加到empInfo字典中。

## 2.6 漂亮打印
如果程序中导入pprint 模块，就可以使用pprint()和pformat()函数，它们将“漂亮打印”一个字典的字。如果想要字典中表项的显示比print()的输出结果更干净，这就有用了。

In [58]:
import pprint
message = "我爱程序设计，我要好好学习，天天向上。"
count = {}
for character in message:
    count.setdefault(character, 0)
    count[character] = count[character] + 1
    
pprint.pprint(count)

{'。': 1,
 '上': 1,
 '习': 1,
 '向': 1,
 '天': 2,
 '好': 2,
 '学': 1,
 '序': 1,
 '我': 2,
 '爱': 1,
 '程': 1,
 '要': 1,
 '计': 1,
 '设': 1,
 '，': 2}


# 3. 字典综合实例：人事统计
现有表格数据如下

|编号|姓名|部门|电子邮箱|电话号码|身份证号码|
|---|----|----|:----|----|----|
|10932|张珊|管理|zhans@163.com|18612345678|520125197907167551|
|10933|李思|软件|lisi@163.com|18612345679|510110195512176250|
|10934|王武|财务|wangwu@example.com|18612345680|510102196510255826|
|10935|赵柳|财务|zhaoliu@163.com|18612345681|510115195810287715|
|10936|钱棋|人事|qianqi@example.com|18612345682|510103196707165650|
|10941|张明|管理|zhangming@example.com|18612345683|510105198112166015|
|10942|赵敏|人事|zhaomin@163.com|18612345684|510105198505225815|
|10945|王红|培训|wanghong@example.com|18612345685|510221196208516555|
|10946|李萧|培训|lixiao@example.com|18612345686|510221196601510810|
|10947|孙科|软件|sunke@163.com|18612345687|510105198507025519|
|10948|刘利|软件|liuli@example.com|18612345688|510227197510280527|

对下列表格进行统计：
* 各部门人数
* 使用163电子邮箱人数
* 性别为男的员工人数

表格中的数据已经存在employeeList列表中：

In [59]:
employeeList = [['10932','张珊','管理','zhans@163.com','18612345678','520125197907167551'],
['10933','李思','软件','lisi@163.com','18612345679','510110195512176250'],
['10934','王武','财务','wangwu@example.com','18612345680','510102196510255826'],
['10935','赵柳','财务','zhaoliu@163.com','18612345681','510115195810287715'],
['10936','钱棋','人事','qianqi@example.com','18612345682','510103196707165650'],
['10941','张明','管理','zhangming@example.com','18612345683','510105198112166015'],
['10942','赵敏','人事','zhaomin@163.com','18612345684','510105198505225815'],
['10945','王红','培训','wanghong@example.com','18612345685','510221196208516555'],
['10946','李萧','培训','lixiao@example.com','18612345686','510221196601510810'],
['10947','孙科','软件','sunke@163.com','18612345687','510105198507025519'],
['10948','刘利','软件','liuli@example.com','18612345688','510227197510280527']]

*** PS. 掌握了如何用程序操作文件后，就不再需要这样读取数据了，可以操作成千上万的数据了，文件操作将在下一部分介绍，期待吧！！！ ***

将employeeList中的数据存储到字典中：

统计各部门人数：

统计使用163电子邮箱人数：

性别为男的员工人数:

# 4. 字典实例2：斐波拉切数列

斐波那契数列（Fibonacci sequence），又称黄金分割数列、因数学家列昂纳多·斐波那契（Leonardoda Fibonacci）。以兔子繁殖为例子而引入，故又称为“兔子数列”，指的是这样一个数列：1、1、2、3、5、8、13、21、34、……

在数学上，斐波纳契数列以如下被以递推的方法定义：F(1)=1，F(2)=1, F(n)=F(n-1)+F(n-2)（n>=3，n∈N*）

这个数列从第3项开始，每一项都等于前两项之和。

* **使用递归的方式实现：**

In [61]:
def fibo(i):
    if i == 0:
        return 0
    elif i == 1:
        return 1
    else:
        return fibo(i-1) + fibo(i-2)

In [63]:
for i in range(10):
    print(fibo(i), end=' ')

0 1 1 2 3 5 8 13 21 34 

试试打印出前50项（运行下面的代码前，做好心理准备:)

In [ ]:
for i in range(50):
    print(fibo(i), end=' ')

<u>如此之慢的原因是什么？</u>
<font color='white'>每次在计算第i项值时，都需要递归调用直到fibo(0)，也就是说像fibo(0),fibo(1),fibo(2),fibo(3)被计算了无数次，如果我们能在第一次计算出来后就存储下来，以供后面使用，会不会快些？<font>

让我们使用字典改进一下：

In [67]:
alreadKnow = {0: 0, 1: 1}

def fibByDic(n):
    if n not in alreadKnow:
        new_value = fib(n-1) + fib(n-2)
        alreadKnow[n] = new_value
    return alreadKnow[n]

In [68]:
for i in range(50):
    print(fibByDic(i), end=' ')

0 1 1 2 3 5 8 13 21 34 55 89 144 233 377 610 987 1597 2584 4181 6765 10946 17711 28657 46368 75025 121393 196418 317811 514229 832040 1346269 2178309 3524578 5702887 9227465 14930352 24157817 39088169 63245986 102334155 165580141 267914296 433494437 701408733 1134903170 1836311903 2971215073 4807526976 7778742049 

和简单的递归相比较，速度是否快到让你怀疑人生？所以，有的时候<b><font size=7 color=red>算法很重要</font><b>。程序设计可以让你的工作由几天节约至几个小时，好的算法可能可以让你的程序运行时间从几个小时节约至几秒钟。

# 5. jieba库的使用

jieba是Python中一个重要的第三方中文分词函数库：

In [72]:
import jieba
jieba.lcut("中国是一个伟大的国家")

Building prefix dict from the default dictionary ...
Dumping model to file cache C:\Users\xiezh\AppData\Local\Temp\jieba.cache
Loading model cost 0.724 seconds.
Prefix dict has been built succesfully.


['中国', '是', '一个', '伟大', '的', '国家']

jieba库是第三方库，不是安装包自带，需要通过pip指令安装：
```Python
pip install jieba   
# 或者 
pip3 install jieba
```

jieba库的解析方法：

|函数|描述|
|---|:---|
|jieba.cut(s)|精确模式，返回一个可迭代的数据类型|
|jieba.cut(s, cut_all=True)|全模式，输出文本s中所有可能单词|
|jieba.cut_for_search(s)|搜索引擎模式，适合搜索引擎建立索引的分词结果|
|jieba.lcut(s)|精确模式，返回一个列表类型，建议使用|
|jieba.lcut(s, cut_all=True)|全模式，返回一个列表类型，建议使用|
|jieba.add_word(w)|向分词词典中增加新词w|

In [74]:
import jieba
jieba.lcut("中华人民共和国是一个伟大的国家")

['中华人民共和国', '是', '一个', '伟大', '的', '国家']

In [75]:
jieba.lcut("中华人民共和国是一个伟大的国家", cut_all=True)

['中华',
 '中华人民',
 '中华人民共和国',
 '华人',
 '人民',
 '人民共和国',
 '共和',
 '共和国',
 '国是',
 '一个',
 '伟大',
 '的',
 '国家']

In [76]:
jieba.lcut_for_search("中华人民共和国是一个伟大的国家")

['中华', '华人', '人民', '共和', '共和国', '中华人民共和国', '是', '一个', '伟大', '的', '国家']

中文切词有时并不是那么容易，比如：

In [77]:
jieba.lcut('南京市长江大桥')

['南京市', '长江大桥']

其实，我想说的是："南京市长"，"江大桥"。这个怎么办？课外了解一下。

## 用jieba做文本词频统计

### 《Hamlet》英文词频统计


In [79]:
def getText():
    txt = open("Documents/hamlet.txt", "r").read()
    txt = txt.lower()
    for ch in '!"#$%&()*+,-./:;<=>?@[\\]^_‘{|}~':
        txt = txt.replace(ch, " ")   #将文本中特殊字符替换为空格
    return txt

hamletTxt = getText()
words  = hamletTxt.split()
counts = {}
for word in words:
    counts[word] = counts.get(word,0) + 1
items = list(counts.items())
items.sort(key=lambda x:x[1], reverse=True) 
for i in range(10):
    word, count = items[i]
    print ("{0:<10}{1:>5}".format(word, count))


the        1138
and         965
to          754
of          669
you         550
i           542
a           542
my          514
hamlet      462
in          436


观察输出结果可以看到，高频单词大多数是冠词、代词、连接词等语法型词汇，并不能代表文章的含义。进一步，可以采用集合类型构建一个排除词汇库excludes，在输出结果中排除这个词汇库中内容。

In [82]:
excludes = {"the","and","of","you","a","i","my","in"}

def getText():
    txt = open("Documents/hamlet.txt", "r").read()
    txt = txt.lower()
    for ch in '!"#$%&()*+,-./:;<=>?@[\\]^_‘{|}~':
        txt = txt.replace(ch, " ")   #将文本中特殊字符替换为空格
    return txt

hamletTxt = getText()
words  = hamletTxt.split()
counts = {}
for word in words:
    counts[word] = counts.get(word,0) + 1
for word in excludes:
    del(counts[word])
items = list(counts.items())
items.sort(key=lambda x:x[1], reverse=True) 
for i in range(10):
    word, count = items[i]
    print ("{0:<10}{1:>5}".format(word, count))

to          754
hamlet      462
it          416
that        391
is          340
not         314
lord        309
his         296
this        295
but         269


### 《三国演义》人物出场统计 

In [84]:
import jieba
txt = open("Documents/三国演义.txt", "r", encoding='utf-8').read()
words  = jieba.lcut(txt)
counts = {}
for word in words:
    if len(word) == 1:  #排除单个字符的分词结果
        continue
    else:
        counts[word] = counts.get(word,0) + 1
items = list(counts.items())
items.sort(key=lambda x:x[1], reverse=True) 
for i in range(15):
    word, count = items[i]
    print ("{0:<10}{1:>5}".format(word, count))

曹操          953
孔明          836
将军          772
却说          656
玄德          585
关公          510
丞相          491
二人          469
不可          440
荆州          425
玄德曰         390
孔明曰         390
不能          384
如此          378
张飞          358


观察输出结果，同一个人物会有不同的名字，这种情况需要整合处理。同时，与英文词频统计类似，需要排除一些人名无关词汇，如“却说”、“将军”等。

In [87]:
import jieba
excludes = {"将军","却说","荆州","二人","不可","不能","如此"}

txt = open("Documents/三国演义.txt", "r", encoding='utf-8').read()
words  = jieba.lcut(txt)
counts = {}

for word in words:
    if len(word) == 1:
        continue
    elif word == "诸葛亮" or word == "孔明曰":
        rword = "孔明"
    elif word == "关公" or word == "云长":
        rword = "关羽"
    elif word == "玄德" or word == "玄德曰":
        rword = "刘备"
    elif word == "孟德" or word == "丞相":
        rword = "曹操"
    else:
        rword = word
    counts[rword] = counts.get(rword,0) + 1
    
for word in excludes:
    del(counts[word])
items = list(counts.items())
items.sort(key=lambda x:x[1], reverse=True) 
for i in range(5):
    word, count = items[i]
    print ("{0:<10}{1:>5}".format(word, count))


曹操         1451
孔明         1383
刘备         1252
关羽          784
张飞          358


请继续完善程序，排除更多无关词汇干扰，总结出场最多的20个人物都有哪些。这里，给出参考答案。

曹操（1451）、孔明（1383）、刘备（1252）、关羽（784）、张飞（358）、
吕布（300）、赵云（278）、孙权（264）、司马懿（221）、周瑜（217）、
袁绍（191）、马超（185）、魏延（180）、黄忠（168）、姜维（151）、
马岱（127）、庞德（122）、孟获（122）、刘表（120）、夏侯惇（116）

# 6.小结
列表和字典是这样的值，它们可以包含多个值，包括其他列表和字典。字典是有用的，因为你可以把一些项（键）映射到另一些项（值），它不像列表，只包含一系列有序的值。

字典中的值是通过方括号访问的，像列表一样。字典不是只能使用整数下标，而是可以用各种数据类型作为键：整型、浮点型、字符串或元组。

通过将程序中的值组织成数据结构，你可以创建真实世界事物的模型。例如一个简单的人事结构。

## 练习
**(1)** 空字典的代码是怎样的？<br>

**(2)** 一个字典包含键'fow'和值42，看起来是怎样的？<br>

**(3)** 字典和列表的主要区别是什么？<br>

**(4)** 如果emp是{'name':'张三', age': 30}，你试图访问emp['gender']，会发生什么？

**(5)** 如果一个字典保存在变量中，表达式'cat' in pets 和'cat' in pets.values()之间的区别是什么？

**(6)** 列表和元组之间的区别是什么？<br>

**(7)** 如果元组中只有一个整数值42，如何输入该元组？<br>

# 作业：《绝地求生》游戏设计

假设你在设计《绝地求生》的游戏。用于对玩家物品清单建模的数据结构是一个字典。其中键是字符串，描述清单中的物品，值是一个整型值，说明玩家有多少该物品。例如，字典值{'绷带': 10, '急救包': 1, '医疗箱': 3, '止痛药': 8, '能量饮料': 3}意味着玩家有10个绷带、1个急救包、3个医疗箱，8瓶止痛药和3瓶能量饮料等。

写一个名为**displayInventory()**的函数，它接受任何可能的物品清单，并显示如下：

<img src="Images/homework61.png" width=90%>

假设游戏玩家可以一次性捡到一个“盒子”(游戏特有说法，可以上网查查看)里所有的物资，这个功能在游戏里是没有的，但是作为游戏总设计师的你决定加上这个功能。盒子使用列表表示，例如['绷带','医疗箱','绷带','绷带' ,'肾上腺素注射器', '能量饮料','肾上腺素注射器']，表示盒子里有3个绷带，1个医疗箱，2个肾上腺素注射器和1瓶能量饮料。

写一个名为**addToInventory(inventory, boxItems)**的函数，其中inventory 参数是一个字典，表示玩家的物品清单（像前面一样），boxItems参数是一个列表，addToInventory()函数应该返回一个字典，表示更新过的物品清单。请注意，列表可以包含多个同样的项。你的代码看起来可能像这样：

<img src="Images/homework62.png" width=90%>

将2个截图作为第5次作业的附件，发送到xzl_swufe@163.com。<br><br>
<b><font color=red>邮件标题必须类似于：作业6_12345678。</font>其中12345678需要替换成你的学号。<br><br><b><font color=red>邮件标题必须类似于：作业6_12345678。</font>其中12345678需要替换成你的学号。<br><br><b><font color=red>邮件标题必须类似于：作业6_12345678。</font>其中12345678需要替换成你的学号。